packages/config/src/discovery/ProjectDiscovery.ts

```diff
@@ -190,90 +188,58 @@ export class ProjectDiscovery {
+      contracts[name] ??= new PermissionedContract(name, address)
+      return contracts[name]
     }
 
     for (const template of templates) {
-      for (const contract of inversion.values()) {
-        const role = contract.roles.find(
-          (r) =>
-            roleNameMatches(template.role.value, r.name) &&
-            r.atName ===
-              (contractOverrides?.[template.role.contract] ??
-                template.role.contract),
+      for (const invertedContract of inversion.values()) {
+        const role = findRoleMatchingTemplate(
+          invertedContract,
+          template,
+          contractOverrides,
         )
         if (!role) {
           continue
         }
 
-        const contractKey = overrides?.[role.name] ?? contract.name ?? role.name
+        const contractKey =
+          overrides?.[role.name] ?? invertedContract.name ?? role.name
 
-        result[contractKey] ??= {
-          name: contractKey,
-          addresses: [EthereumAddress(contract.address)],
-          contractDescription: {},
-          taggedNames: {},
-        }
-        const entry = result[contractKey]
+        const contractAddress = EthereumAddress(invertedContract.address)
+        const contract = getContract(contractKey, contractAddress)
+        const referenced = getContract(role.atName, role.atAddress)
 
         if (template.description !== undefined) {
-          entry.contractDescription[role.name] ??= []
-          entry.contractDescription[role.name].push(
+          contract.addDescription(
             stringFormat(template.description, role.atName),
           )
         }
 
         if (template.tags !== undefined) {
           for (const tag of template.tags) {
-            entry.taggedNames[tag] ??= []
-            entry.taggedNames[tag].push(role.atName)
+            contract.addTag(tag, role.atName)
+            referenced.addTagReference(tag, contractKey)
           }
         }
       }
     }
 
-    return Object.values(result).map((entry) => ({
-      name: entry.name,
-      accounts: entry.addresses.map((a) => this.formatPermissionedAccount(a)),
-      description: Object.values(entry.contractDescription)
-        .flat()
-        .concat(
-          Object.entries(entry.taggedNames).map(([tag, contracts]) =>
-            stringFormat(
-              StackPermissionsTagDescription[tag as StackPermissionsTag],
-              contracts.join(', '),
-            ),
-          ),
-        )
-        .join(' '),
-      chain: this.chain,
-    }))
+    return Object.values(contracts)
+      .map((contract) => {
+        const description = contract.generateDescription()
+        if (description !== '') {
+          return {
+            name: contract.name,
+            accounts: [this.formatPermissionedAccount(contract.address)],
+            description,
+            chain: this.chain,
+          }
+        }
+      })
+      .filter((x) => !!x) as ScalingProjectPermission[]
```

Comment:
I don't like this operator, it's to broad. Wouldn't it be better to return undefined above in the else clause and later filter undefined? In addition, I feel like this operator makes the code harder to read/reason about the possibilities than can happen.

We have this util function (`packages/shared-pure/src/utils/notUndefined.ts`), wdyt about using it?
```typescript
export function notUndefined<T>(value: T | undefined): value is T {
  return value !== undefined
}
```

Reply:
nice, I'll use it.